Low cost DIY modem using Unet audio (#66)
org-arl/unet-blog: _posts/2022-09-20-Project-Sabine-Low-cost-DIY-underwater-modem-using-COTS-components-and-Unet-audio.md

---
layout: post
comments: true
title: Low-cost DIY underwater modem using COTS components and Unet audio
date: 20/9/2022
author: Manu Ignatius, Sukanta K. Hazra
categories: howto
feature-img: "assets/img/stdma/stdma-unet.jpg"
thumbnail: "assets/img/stdma/stdma-unet.jpg"
tags: [howto, unetaudio, unetstack, modems, phy, janus, sabine]
---

## Introduction

> In the article titled [Converting your laptop into a JANUS modem using Unet audio](https://blog.unetstack.net/converting-your-laptop-into-a-janus-modem-using-unetaudio), we saw how UnetStack, a Software-Defined Open Architecture Modem (SDOAM), can be used to turn your laptop into an acoustic modem. A true SDOAM can easily be extended to run on various hardware platforms with different types of physical layers (e.g. Acoustic, optical, RF) with minimal efforts to address different use cases. In this article, we will show how you can build a low-cost (< USD 300) , DIY underwater acoustic modem using only Commercial Off-The-Shelf (COTS) components and Unet audio. The goal of this article is to demonstrate the approach and basic steps one can use to build a low cost acoustic modem using Unet audio. We encourage the user to build on these principles to build create versions of the DIY modem.

Reviewer: Suggestions:
- `user` => `reader`
- `to build create` => `to build your own`

> In the article titled [Converting your laptop into a JANUS modem using Unet audio](https://blog.unetstack.net/converting-your-laptop-into-a-janus-modem-using-unetaudio), we saw how UnetStack, a Software-Defined Open Architecture Modem (SDOAM), can be used to turn your laptop into an acoustic modem. A true SDOAM can easily be extended to run on various hardware platforms with different types of physical layers (e.g. Acoustic, optical, RF) with minimal efforts to address different use cases. In this article, we will show how you can build a low-cost (< USD 300) , DIY underwater acoustic modem using only Commercial Off-The-Shelf (COTS) components and Unet audio. The goal of this article is to demonstrate the approach and basic steps one can use to build a low cost acoustic modem using Unet audio. We encourage the user to build on these principles to build create versions of the DIY modem.

Author: ```suggestion
In the article titled [Converting your laptop into a JANUS modem using Unet audio](https://blog.unetstack.net/converting-your-laptop-into-a-janus-modem-using-unetaudio), we saw how UnetStack, a Software-Defined Open Architecture Modem (SDOAM), can be used to turn your laptop into an acoustic modem. A true SDOAM can easily be extended to run on various hardware platforms with different types of physical layers (e.g. Acoustic, optical, RF) with minimal efforts to address different use cases. In this article, we will show how you can build a low-cost (< USD 300) , DIY underwater acoustic modem using only Commercial Off-The-Shelf (COTS) components and Unet audio. The goal of this article is to demonstrate the approach and basic steps one can use to build a low cost acoustic modem using Unet audio. We encourage the reader to build on these principles to build your own versions of the DIY modem.
```

## Motivation and Approach

> Most commercially available modems are beyond the reach of a technology enthusiast or a hobbyist underwater roboticist, due to their price point. However, applications such as DIY underwater vehicles (e.g. for [The Singapore AUV Challenge](https://sauvc.org/)), low-cost swarms, etc. need low-cost underwater modems to realize their true potential. While there have been projects like the [AHOI modem](https://dl.acm.org/doi/10.1145/3376921), [SeaModem](https://ieeexplore.ieee.org/document/7271721), [Seatrac](https://ieeexplore.ieee.org/document/7271578), all of them has a hardware component associated with it. This means the user needs to spend time building the hardware, which can be cumbersome and time-consuming. Adding additional hardware to a micro-AUV can be challenging due to space and power limitations. Ensuring long-term support for these can also be tricky, depending on the creator's other commitments.

Reviewer: - `all of them has a hardware component associated with it` => `all of them have custom hardware associated with them`

> Most commercially available modems are beyond the reach of a technology enthusiast or a hobbyist underwater roboticist, due to their price point. However, applications such as DIY underwater vehicles (e.g. for [The Singapore AUV Challenge](https://sauvc.org/)), low-cost swarms, etc. need low-cost underwater modems to realize their true potential. While there have been projects like the [AHOI modem](https://dl.acm.org/doi/10.1145/3376921), [SeaModem](https://ieeexplore.ieee.org/document/7271721), [Seatrac](https://ieeexplore.ieee.org/document/7271578), all of them has a hardware component associated with it. This means the user needs to spend time building the hardware, which can be cumbersome and time-consuming. Adding additional hardware to a micro-AUV can be challenging due to space and power limitations. Ensuring long-term support for these can also be tricky, depending on the creator's other commitments.

Author: ```suggestion
Most commercially available modems are beyond the reach of a technology enthusiast or a hobbyist underwater roboticist, due to their price point. However, applications such as DIY underwater vehicles (e.g. for [The Singapore AUV Challenge](https://sauvc.org/)), low-cost swarms, etc. need low-cost underwater modems to realize their true potential. While there have been projects like the [AHOI modem](https://dl.acm.org/doi/10.1145/3376921), [SeaModem](https://ieeexplore.ieee.org/document/7271721), [Seatrac](https://ieeexplore.ieee.org/document/7271578), all of them have custom hardware associated with them. This means the user needs to spend time building the hardware, which can be cumbersome and time-consuming. Adding additional hardware to a micro-AUV can be challenging due to space and power limitations. Ensuring long-term support for these can also be tricky, depending on the creator's other commitments.
```

These are some of the challenges for anyone looking to use one of the existing projects to build on.

> Unet audio being a true SDOAM addresses these by using a commonly available hardware platform such as Raspberry Pi and using its sound card to transmit and receive using an underwater speaker and mic. This approach allows the user to run Unet audio in their single board computer (SBC) thereby avoiding the need for dedicated digital hardware (e.g. SBC, FPGA). The audio amplifiers can then be chosen based on the use case. By using Unet audio, the user gets access to all the developer tools and the extendability to build their protocols on top of the DIY modem. Long term support is guaranteed since UnetStack is actively developed by both academia ([ARL](http://arl.nus.edu.sg/)) and industry ([Subnero](https://subnero.com/)).

Reviewer 2: ```suggestion
Unet audio being a true SDOAM addresses these by using a commonly available hardware platform such as Raspberry Pi and using its sound card to transmit and receive using an underwater speaker and microphone. This approach allows the user to run Unet audio in their single board computer (SBC) thereby avoiding the need for dedicated digital hardware (e.g. SBC, FPGA). The audio amplifiers can then be chosen based on the use case. By using Unet audio, the user gets access to all the developer tools and the extendability to build their protocols on top of the DIY modem. Long term support is guaranteed since UnetStack is actively developed by both academia ([ARL](http://arl.nus.edu.sg/)) and industry ([Subnero](https://subnero.com/)).
```

## Overview of the DIY modem

A modem has three main functions

> 1. Convert user-given data to signals and transmit it to the receiver.

Reviewer: ```suggestion
1. Convert user-given data to a signal and transmit it to the receiver.
```

> 2. Decode the signals received at the receiver to recover the original data.

Reviewer: ```suggestion
2. Decode the signal received at the receiver to recover the original data.
```

So the main hardware components of an SDOAM will be an SBC to run Unet audio, a sound card (with speaker out and mic in), audio amplifiers, and speaker and mic. The following sections list the hardware chain for both the transmit and receive chains.

### Transmit chain

Connect the speaker output from the sound card (either a discrete or USB sound card connected to an SBC) to the amplifier module input. The amplifier output is then connected to the underwater speaker. Depending on the amplifier used, it can be powered using the USB output of the RPi or a 12V battery.

<p align="center"><img width="70%" src="../assets/img/sabine/txc.png"/></p>
<p align="center"><em>Transmit chain</em></p>

### Receive chain

For the receiver side, the underwater microphone is connected directly to the mic in of the soundcard of the SBC.

<p align="center"><img width="70%" src="../assets/img/sabine/rxc.png"/></p>
<p align="center"><em>Receive chain</em></p>

## Hardware Selection

### Digital Hardware

> One of the most important hardware components of an SDOAM is the digital hardware that runs Unet audio. We have selected a Raspberry Pi (3b+) due to its popularity, low cost, ease of availability, and community support to be the digital hardware to run Unet audio. This can easily be swapped with any similar SBCs that have an arm processor running Linux on it. A USB sound card connected to the Raspberry Pi is used as the data acquisition system.

Reviewer 2: Can we give a link for what kind of USB Sound card worked for us?

> One of the most important hardware components of an SDOAM is the digital hardware that runs Unet audio. We have selected a Raspberry Pi (3b+) due to its popularity, low cost, ease of availability, and community support to be the digital hardware to run Unet audio. This can easily be swapped with any similar SBCs that have an arm processor running Linux on it. A USB sound card connected to the Raspberry Pi is used as the data acquisition system.

Author: ```suggestion
One of the most important hardware components of an SDOAM is the digital hardware that runs Unet audio. We have selected a Raspberry Pi (3b+) due to its popularity, low cost, ease of availability, and community support to be the digital hardware to run Unet audio. This can easily be swapped with any similar SBCs that have an arm processor running Linux on it. A [USB sound card](https://www.aliexpress.com/item/4000598865391.html) connected to the Raspberry Pi is used as the data acquisition system.
```

2. Power on your Raspberry Pi
3. Ensure that the USB Audio Card is detected using `aplay -l`

```
> aplay -l
card 0: ...
card 1: Audio [USB Audio], device 0: USB Audio [USB Audio]
  Subdevices: 0/1
  Subdevice #0: subdevice #0
```
4. Disable the RaspberyPi onboard sound card, by commenting out the line `dtparam=audio=on` in `/boot/config.txt`

```
> sudo nano /boot/config.txt

# Enable audio (loads snd_bcm2835)
# dtparam=audio=on
```

5. Make the USB Audio card the default device in alsa by updating `defaults.ctl.card` and `defaults.pcm.card` to `1` in `/usr/share/alsa/alsa.conf`

```
defaults.ctl.card 1
defaults.pcm.card 1
```

6. Reboot your Raspberry Pi and verify that `card 0:` is not enumerated by `aplay -l`

```
> aplay -l
card 1: Audio [USB Audio], device 0: USB Audio [USB Audio]
  Subdevices: 0/1
  Subdevice #0: subdevice #0
```

7. You can test if your audio setup works properly using  `speaker-test -c2`

### Dependencies

1. Install Java8

```sh
wget https://github.com/AdoptOpenJDK/openjdk8-binaries/releases/download/jdk8u265-b01/OpenJDK8U-jdk_arm_linux_hotspot_8u265b01.tar.gz
tar xzf OpenJDK8U-jdk_arm_linux_hotspot_8u265b01.tar.gz
rm jdk8u265-b01/src.zip

> sudo mkdir -p /usr/lib/jvm/java-8-openjdk-armhf

Reviewer 2: Why are we using `armhf`? That's for 32bit ARM. I am guessing it will work because of backward compatibility, but maybe easier to use `arm64` or `aarch64` for longevity reasons.

Author: I copied these instructions from the document you provided. Perhaps we can test this with the next release in rpi and update later? These are the instructions that worked in the current setup.
